File: frontend/src/api/workflow.ts
import type {
  IntakeFormPayload,
  MaterialSummary,
  ObservationDetail,
  PaperSummary,
  ProblemDetail,
  PropertyOption,
  TermOption,
  VerificationStatus,
} from '../types/workflow'

const API_BASE = import.meta.env.VITE_API_BASE_URL ?? '/api'

export class ApiError extends Error {
  status: number
  problem?: ProblemDetail

  constructor(message: string, status: number, problem?: ProblemDetail) {
    super(message)
    this.name = 'ApiError'
    this.status = status
    this.problem = problem
  }
}

async function apiFetch<T>(path: string, init?: RequestInit, token?: string): Promise<T> {
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    ...(init?.headers as Record<string, string> | undefined),
  }
  if (token?.trim()) {
    headers['Authorization'] = `Bearer ${token.trim()}`
  }

  const response = await fetch(`${API_BASE}${path}`, {
    ...init,
    headers,
  })

  if (!response.ok) {
    let problem: ProblemDetail | undefined
    let message = `请求失败（${response.status}）`
    try {
      const data = await response.json()
      problem = data as ProblemDetail
      if (problem.detail) {
        message = `${problem.title ? problem.title + ': ' : ''}${problem.detail}`
      }
    } catch {
      // 无法解析 JSON
    }
    throw new ApiError(message, response.status, problem)
  }

  return response.json() as Promise<T>
}

export async function fetchPapers(): Promise<PaperSummary[]> {
  const res = await apiFetch<{ items: PaperSummary[] }>('/v1/papers?limit=100')
  return res.items
}

export async function fetchMaterials(): Promise<MaterialSummary[]> {
  const res = await apiFetch<{ items: MaterialSummary[] }>('/v1/materials?limit=100')
  return res.items
}

export async function fetchProperties(): Promise<PropertyOption[]> {
  return apiFetch<PropertyOption[]>('/v1/properties')
}

export async function fetchTerms(namespace: string): Promise<TermOption[]> {
  return apiFetch<TermOption[]>(`/v1/workflow/terms?namespace=${encodeURIComponent(namespace)}`)
}

export async function fetchObservationDetail(observationId: string): Promise<ObservationDetail> {
  return apiFetch<ObservationDetail>(`/v1/workflow/observations/${observationId}`)
}

export async function submitIntake(
  form: IntakeFormPayload,
  idempotencyKey: string,
  reviewerToken: string
): Promise<{ observation_id: string; row_version: number }> {
  const numericVal = parseFloat(form.originalValue)
  const isNum = !isNaN(numericVal)

  const body: Record<string, unknown> = {
    document: {
      storage_uri: form.docStorageUri.trim(),
      sha256: form.docSha256.trim().toLowerCase(),
      document_type: form.docType || 'main_article',
    },
    sample: {
      sample_label: form.sampleLabel.trim(),
      sample_type_term_id: form.sampleTypeTermId,
      thickness_value: form.sampleThickness ? parseFloat(form.sampleThickness) : null,
      thickness_unit: form.sampleThicknessUnit.trim() || null,
      substrate_material: form.sampleSubstrate.trim() || null,
    },
    evidence: {
      page_number: parseInt(form.evdPage, 10),
      section: form.evdSection.trim() || null,
      figure_number: form.evdFigure.trim() || null,
      table_number: form.evdTable.trim() || null,
      text_snippet: form.evdText.trim(),
      fragment_type: 'paragraph',
    },
    measurement: {
      measurement_type_term_id: form.measTypeTermId,
      instrument: form.measInstrument.trim() || null,
      temperature_value: form.measTemp ? parseFloat(form.measTemp) : null,
      temperature_unit: form.measTempUnit.trim() || null,
    },
    observation: {
      property_definition_id: form.propertyId,
      value_kind: 'scalar',
      value_numeric: isNum ? numericVal : null,
      original_value_text: form.originalValue.trim(),
      normalized_value: null,
      normalized_unit_term_id: null,
      uncertainty_lower: form.uncertaintyLower ? parseFloat(form.uncertaintyLower) : null,
      uncertainty_upper: form.uncertaintyUpper ? parseFloat(form.uncertaintyUpper) : null,
      condition_temperature_value: form.measTemp ? parseFloat(form.measTemp) : null,
      condition_temperature_unit: form.measTempUnit.trim() || null,
      verification_status: 'HUMAN_REVIEWED',
    },
  }

  if (form.paperMode === 'existing') {
    body.paper_id = form.existingPaperId
  } else {
    body.paper = {
      title: form.newPaperTitle.trim(),
      doi: form.newPaperDoi.trim() || null,
      journal: form.newPaperJournal.trim() || null,
      publication_year: form.newPaperYear ? parseInt(form.newPaperYear, 10) : null,
    }
  }

  if (form.materialMode === 'existing') {
    body.material_id = form.existingMaterialId
  } else {
    body.material = {
      canonical_formula: form.newMaterialFormula.trim(),
      reduced_formula: form.newMaterialFormula.trim(),
      chemical_system: form.newMaterialSystem.trim(),
      name: form.newMaterialName.trim() || null,
      aliases: [],
    }
  }

  return apiFetch<{ observation_id: string; row_version: number }>(
    '/v1/workflow/intake',
    {
      method: 'POST',
      headers: {
        'Idempotency-Key': idempotencyKey,
      },
      body: JSON.stringify(body),
    },
    reviewerToken
  )
}

export async function submitReview(
  observationId: string,
  decision: VerificationStatus,
  comment: string | null,
  reviewer: string,
  rowVersion: number,
  reviewerToken: string
): Promise<{ observation_id: string; new_status: VerificationStatus; row_version: number }> {
  return apiFetch<{ observation_id: string; new_status: VerificationStatus; row_version: number }>(
    `/v1/workflow/observations/${observationId}/review`,
    {
      method: 'POST',
      headers: {
        'If-Match': `W/"${rowVersion}"`,
      },
      body: JSON.stringify({
        decision,
        comment: comment?.trim() || null,
        reviewer: reviewer.trim(),
      }),
    },
    reviewerToken
  )
}

File: app/models/workflow.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import Field, field_validator, model_validator

from app.domain.workflow import validate_storage_uri
from app.models.common import APIModel, ValueKind, VerificationStatus
from app.models.literature import PaperCreate
from app.models.material import MaterialCreate


class DocumentIntake(APIModel):
    storage_uri: str = Field(min_length=5, max_length=1024)
    sha256: str = Field(min_length=64, max_length=64)
    document_type: str = Field(default="main_article", max_length=32)

    @field_validator("storage_uri")
    @classmethod
    def check_storage_uri(cls, v: str) -> str:
        try:
            return validate_storage_uri(v)
        except Exception as e:
            raise ValueError(str(e)) from e


class SampleIntake(APIModel):
    sample_label: str = Field(min_length=1, max_length=255)
    sample_type_term_id: UUID
    thickness_value: float | None = Field(default=None, ge=0)
    thickness_unit: str | None = Field(default=None, max_length=32)
    substrate_material: str | None = Field(default=None, max_length=255)
    description: str | None = None


class EvidenceIntake(APIModel):
    page_number: int = Field(ge=1)
    section: str | None = Field(default=None, max_length=512)
    figure_number: str | None = Field(default=None, max_length=64)
    table_number: str | None = Field(default=None, max_length=64)
    text_snippet: str = Field(min_length=1)
    fragment_type: str = Field(default="paragraph", max_length=32)


class MeasurementIntake(APIModel):
    measurement_type_term_id: UUID
    instrument: str | None = Field(default=None, max_length=255)
    temperature_value: float | None = None
    temperature_unit: str | None = Field(default=None, max_length=32)
    heating_rate_value: float | None = None
    heating_rate_unit: str | None = Field(default=None, max_length=32)


class ObservationIntake(APIModel):
    property_definition_id: UUID
    value_kind: ValueKind = ValueKind.SCALAR
    value_numeric: Decimal | float | None = None
    value_min: Decimal | float | None = None
    value_max: Decimal | float | None = None
    value_text: str | None = None
    value_boolean: bool | None = None

    original_value_text: str = Field(min_length=1, max_length=255)
    original_unit_text: str = Field(min_length=1, max_length=64)
    normalized_value: Decimal | float | None = None
    normalized_unit_term_id: UUID | None = None

    uncertainty_lower: Decimal | float | None = None
    uncertainty_upper: Decimal | float | None = None
    condition_temperature_value: Decimal | float | None = None
    condition_temperature_unit: str | None = Field(default=None, max_length=32)
    quality_score: float | None = Field(default=None, ge=0, le=1)
    verification_status: VerificationStatus = VerificationStatus.HUMAN_REVIEWED

    @field_validator("verification_status")
    @classmethod
    def validate_intake_status(cls, v: VerificationStatus) -> VerificationStatus:
        if v in (VerificationStatus.AI_EXTRACTED, VerificationStatus.AI_VALIDATED):
            raise ValueError(
                f"人工登记接口不允许直接录入 {v.value} 状态。"
                "AI 提取结果必须先进入暂存区 (/v1/extractions/candidates) 经审核后晋升"
            )
        if v != VerificationStatus.HUMAN_REVIEWED:
            raise ValueError("人工登记接口初始状态必须为 HUMAN_REVIEWED")
        return v

    @model_validator(mode="after")
    def validate_normalization_fields(self) -> ObservationIntake:
        if self.normalized_value is not None and self.normalized_unit_term_id is None:
            raise ValueError("提供 normalized_value 时必须同时提供 normalized_unit_term_id")
        if self.normalized_unit_term_id is not None and self.normalized_value is None:
            raise ValueError("提供 normalized_unit_term_id 时必须同时提供 normalized_value")
        return self

    @model_validator(mode="after")
    def validate_value(self) -> ObservationIntake:
        if self.value_kind == ValueKind.SCALAR:
            if self.value_numeric is None and self.normalized_value is None:
                raise ValueError("标量观测必须提供 value_numeric 或 normalized_value")
        elif self.value_kind == ValueKind.RANGE:
            if self.value_min is None or self.value_max is None:
                raise ValueError("范围观测必须提供 value_min 与 value_max")
            if self.value_min > self.value_max:
                raise ValueError("value_min 不能大于 value_max")
        elif self.value_kind in (ValueKind.TEXT, ValueKind.CATEGORICAL):
            if not self.value_text:
                raise ValueError("文本/分类型观测必须提供 value_text")
        elif self.value_kind == ValueKind.BOOLEAN:
            if self.value_boolean is None:
                raise ValueError("布尔型观测必须提供 value_boolean")
        return self


class WorkflowIntakeRequest(APIModel):
    # 文献：严格二选一（paper_id 与 paper 必须且只能提供其中之一）
    paper_id: UUID | None = None
    paper: PaperCreate | None = None

    # 文档元数据
    document: DocumentIntake

    # 材料：严格二选一（material_id 与 material 必须且只能提供其中之一）
    material_id: UUID | None = None
    material: MaterialCreate | None = None

    # 实体样品
    sample: SampleIntake

    # 证据片段
    evidence: EvidenceIntake

    # 实验测量
    measurement: MeasurementIntake

    # 科学观测
    observation: ObservationIntake

    @model_validator(mode="after")
    def validate_entities(self) -> WorkflowIntakeRequest:
        if (self.paper_id is None and self.paper is None) or (
            self.paper_id is not None and self.paper is not None
        ):
            raise ValueError("paper_id 与 paper 必须二选一（不能同时提供或同时为空）")
        if (self.material_id is None and self.material is None) or (
            self.material_id is not None and self.material is not None
        ):
            raise ValueError("material_id 与 material 必须二选一（不能同时提供或同时为空）")
        return self


class WorkflowIntakeResponse(APIModel):
    observation_id: UUID
    paper_id: UUID
    document_id: UUID
    artifact_id: UUID
    material_id: UUID
    sample_id: UUID
    measurement_id: UUID
    evidence_id: UUID
    verification_status: VerificationStatus
    row_version: int
    created_at: datetime


class ObservationDetailRead(APIModel):
    id: UUID
    verification_status: VerificationStatus
    row_version: int
    created_at: datetime
    updated_at: datetime

    value_kind: str
    value_numeric: float | None = None
    value_min: float | None = None
    value_max: float | None = None
    value_text: str | None = None
    value_boolean: bool | None = None
    original_value_text: str | None = None
    original_unit_text: str | None = None
    normalized_value: float | None = None
    normalized_unit: str | None = None

    uncertainty_lower: float | None = None
    uncertainty_upper: float | None = None
    condition_temperature_value: float | None = None
    condition_temperature_unit: str | None = None
    quality_score: float | None = None

    property: dict
    material: dict
    sample: dict
    measurement: dict | None = None
    evidence: list[dict] = Field(default_factory=list)


class ObservationReviewRequest(APIModel):
    decision: VerificationStatus
    comment: str | None = Field(default=None, max_length=2000)
    reviewer: str = Field(min_length=1, max_length=255)


class ObservationReviewResponse(APIModel):
    observation_id: UUID
    previous_status: VerificationStatus
    new_status: VerificationStatus
    row_version: int
    reviewer: str
    decision: VerificationStatus
    comment: str | None = None
    reviewed_at: datetime
